src/daft-scan/src/scan_task_iters.rs

```diff
@@ -177,137 +178,225 @@ impl<'a> Iterator for MergeByFileSize<'a> {
     }
 }
 
-#[must_use]
-pub(crate) fn split_by_row_groups(
-    scan_tasks: BoxScanTaskIter,
-    max_tasks: usize,
-    min_size_bytes: usize,
-    max_size_bytes: usize,
-) -> BoxScanTaskIter {
-    let mut scan_tasks = itertools::peek_nth(scan_tasks);
-
-    // only split if we have a small amount of files
-    if scan_tasks.peek_nth(max_tasks).is_some() {
-        Box::new(scan_tasks)
-    } else {
-        Box::new(
-            scan_tasks
-                .map(move |t| -> DaftResult<BoxScanTaskIter> {
-                    let t = t?;
-
-                    /* Only split parquet tasks if they:
-                        - have one source
-                        - use native storage config
-                        - have no specified chunk spec or number of rows
-                        - have size past split threshold
-                        - no iceberg delete files
-                    */
-                    if let (
-                        FileFormatConfig::Parquet(ParquetSourceConfig {
-                            field_id_mapping, ..
-                        }),
-                        StorageConfig::Native(_),
-                        [source],
-                        Some(None),
-                        None,
-                    ) = (
-                        t.file_format_config.as_ref(),
-                        t.storage_config.as_ref(),
-                        &t.sources[..],
-                        t.sources.first().map(DataSource::get_chunk_spec),
-                        t.pushdowns.limit,
-                    ) && source
-                        .get_size_bytes()
-                        .map_or(true, |s| s > max_size_bytes as u64)
-                      && source
-                        .get_iceberg_delete_files()
-                        .map_or(true, std::vec::Vec::is_empty)
-                    {
-                        let (io_runtime, io_client) =
-                            t.storage_config.get_io_client_and_runtime()?;
+#[derive(Default)]
+struct SplitParquetByRowGroupsAccumulator {
+    row_group_indices: Vec<usize>,
+    size_bytes: f64,
+    num_rows: usize,
+}
 
-                        let path = source.get_path();
+impl SplitParquetByRowGroupsAccumulator {
+    fn accumulate(
+        &mut self,
+        rg_idx: &usize,
+        ctx: &SplitParquetByRowGroupsContext,
+    ) -> Option<DataSource> {
+        let rg = ctx.file_metadata.row_groups.get(rg_idx).unwrap();
+
+        // Estimate the materialized size of this rowgroup and add it to curr_size_bytes
+        self.size_bytes += ctx
+            .scantask_estimated_size_bytes
+            .map_or(0., |est_materialized_size| {
+                (rg.compressed_size() as f64 / ctx.total_rg_compressed_size() as f64)
+                    * est_materialized_size as f64
+            });
+        self.num_rows += rg.num_rows();
+        self.row_group_indices.push(*rg_idx);
+
+        // Flush the accumulator if necessary
```

Comment:
This logic has been changed from the past:

1. We use `ScanTask::estimated_size_bytes` when adding to the `self.size_bytes` accumulator, instead of the size of the file on disk. This should give us a "more accurate" sizing than before, since it should account for things such as column pruning (ideally).

2. When `ScanTask::estimated_size_bytes` is not provided, we just always flush (i.e. every rowgroup becomes its own ScanTask). Alternatively, we can just avoid splitting at all. Not sure what the intended behavior there should be.